File: packages/native-helpers/linux-helper-py/linux_helper/handlers/accessibility.py
"""Accessibility context via GNOME Shell D-Bus."""

import json
import os
import subprocess
import sys
import time


def _get_gnome_active_window() -> dict:
    default = {"appName": None, "windowTitle": None, "pid": 0}

    try:
        script = """
(function() {
  let w = global.display.get_focus_window();
  if (!w) return JSON.stringify({title: null, wmClass: null, pid: 0});
  return JSON.stringify({
    title: w.get_title(),
    wmClass: w.get_wm_class(),
    pid: w.get_pid()
  });
})()
""".strip()

        result = subprocess.run(
            [
                "gdbus", "call", "--session",
                "--dest", "org.gnome.Shell",
                "--object-path", "/org/gnome/Shell",
                "--method", "org.gnome.Shell.Eval",
                script,
            ],
            capture_output=True, text=True, timeout=3,
        )

        # Parse gdbus output: (true, '{"title":"...","wmClass":"...","pid":123}')
        stdout = result.stdout.strip()
        if "(true," not in stdout:
            return default

        # Extract JSON string between single quotes
        start = stdout.index("'") + 1
        end = stdout.rindex("'")
        parsed = json.loads(stdout[start:end])

        return {
            "appName": parsed.get("wmClass"),
            "windowTitle": parsed.get("title"),
            "pid": parsed.get("pid", 0),
        }
    except Exception:
        return default


def handle_get_accessibility_context(params: dict) -> dict:
    info = _get_gnome_active_window()

    if not info["appName"] and not info["windowTitle"]:
        return {"context": None}

    return {
        "context": {
            "schemaVersion": "2.0",
            "application": {
                "name": info["appName"],
                "bundleIdentifier": None,
                "version": None,
                "pid": info["pid"],
            },
            "windowInfo": {
                "title": info["windowTitle"],
                "url": None,
            },
            "focusedElement": None,
            "textSelection": None,
            "timestamp": int(time.time()),
            "metrics": {
                "totalTimeMs": 0,
                "textMarkerAttempted": False,
                "textMarkerSucceeded": False,
                "fallbacksUsed": [],
                "errors": [],
                "timedOut": False,
                "webAreaRetryAttempted": False,
                "webAreaFound": False,
                "webAreaRetrySucceeded": False,
            },
        }
    }


def _check_command(cmd: str) -> bool:
    try:
        subprocess.run(["which", cmd], capture_output=True, check=True)
        return True
    except (subprocess.SubprocessError, OSError):
        return False


def _check_evdev_access() -> bool:
    try:
        for entry in os.listdir("/dev/input"):
            if entry.startswith("event"):
                if os.access(f"/dev/input/{entry}", os.R_OK):
                    return True
    except OSError:
        pass
    return False


def handle_get_accessibility_status(params: dict) -> dict:
    all_ok = all([
        _check_command("wl-copy"),
        _check_command("ydotool"),
        _check_command("pactl"),
        _check_evdev_access(),
    ])
    return {"hasPermission": all_ok, "isEnabled": True}


def handle_request_accessibility_permission(params: dict) -> dict:
    status = handle_get_accessibility_status({})
    return {"granted": status["hasPermission"]}


def handle_get_accessibility_tree_details(params: dict) -> dict:
    return {"tree": None}

File: packages/native-helpers/linux-helper-ts/src/handlers/accessibility.ts
import { execFile } from "node:child_process";
import { promisify } from "node:util";
import * as fs from "node:fs";

const execFileAsync = promisify(execFile);

/**
 * Get active window info from GNOME Shell via D-Bus eval.
 */
async function getGnomeActiveWindow(): Promise<{
  appName: string | null;
  windowTitle: string | null;
  pid: number;
}> {
  const defaultResult = { appName: null, windowTitle: null, pid: 0 };

  try {
    // Get window title and WM class via GNOME Shell eval
    const script = `
      (function() {
        let w = global.display.get_focus_window();
        if (!w) return JSON.stringify({title: null, wmClass: null, pid: 0});
        return JSON.stringify({
          title: w.get_title(),
          wmClass: w.get_wm_class(),
          pid: w.get_pid()
        });
      })()
    `.trim();

    const { stdout } = await execFileAsync(
      "gdbus",
      [
        "call",
        "--session",
        "--dest",
        "org.gnome.Shell",
        "--object-path",
        "/org/gnome/Shell",
        "--method",
        "org.gnome.Shell.Eval",
        script,
      ],
      { timeout: 3000 },
    );

    // gdbus returns: (true, '{"title":"...","wmClass":"...","pid":123}')
    const match = stdout.match(/\(true,\s*'(.+)'\)/s);
    if (!match) return defaultResult;

    const parsed = JSON.parse(match[1]);
    return {
      appName: parsed.wmClass ?? null,
      windowTitle: parsed.title ?? null,
      pid: parsed.pid ?? 0,
    };
  } catch {
    return defaultResult;
  }
}

export async function handleGetAccessibilityContext(
  _params: Record<string, unknown>,
): Promise<{ context: Record<string, unknown> | null }> {
  const windowInfo = await getGnomeActiveWindow();

  if (!windowInfo.appName && !windowInfo.windowTitle) {
    return { context: null };
  }

  return {
    context: {
      schemaVersion: "2.0",
      application: {
        name: windowInfo.appName,
        bundleIdentifier: null,
        version: null,
        pid: windowInfo.pid,
      },
      windowInfo: {
        title: windowInfo.windowTitle,
        url: null,
      },
      focusedElement: null,
      textSelection: null,
      timestamp: Math.floor(Date.now() / 1000),
      metrics: {
        totalTimeMs: 0,
        textMarkerAttempted: false,
        textMarkerSucceeded: false,
        fallbacksUsed: [],
        errors: [],
        timedOut: false,
        webAreaRetryAttempted: false,
        webAreaFound: false,
        webAreaRetrySucceeded: false,
      },
    },
  };
}

export async function handleGetAccessibilityStatus(
  _params: Record<string, unknown>,
): Promise<{ hasPermission: boolean; isEnabled: boolean }> {
  const checks = await Promise.all([
    checkCommand("wl-copy"),
    checkCommand("ydotool"),
    checkCommand("pactl"),
    checkEvdevAccess(),
  ]);

  const allAvailable = checks.every(Boolean);
  return { hasPermission: allAvailable, isEnabled: true };
}

export async function handleRequestAccessibilityPermission(
  _params: Record<string, unknown>,
): Promise<{ granted: boolean }> {
  // On Linux, permissions are managed via group membership
  // We can't programmatically grant them
  const status = await handleGetAccessibilityStatus({});
  return { granted: status.hasPermission };
}

export async function handleGetAccessibilityTreeDetails(
  _params: Record<string, unknown>,
): Promise<{ tree: null }> {
  // Stub - AT-SPI2 integration deferred
  return { tree: null };
}

async function checkCommand(cmd: string): Promise<boolean> {
  try {
    await execFileAsync("which", [cmd]);
    return true;
  } catch {
    return false;
  }
}

async function checkEvdevAccess(): Promise<boolean> {
  try {
    const entries = fs.readdirSync("/dev/input");
    for (const entry of entries) {
      if (entry.startsWith("event")) {
        try {
          fs.accessSync(`/dev/input/${entry}`, fs.constants.R_OK);
          return true;
        } catch {
          continue;
        }
      }
    }
    return false;
  } catch {
    return false;
  }
}
